res/controllers/Traktor-Kontrol-S4-MK3.js

```diff
@@ -0,0 +1,3066 @@
+            engine.setValue(this.group, "loaded_chain_preset", presetNumber + 1);
+            this.mixer.firstPressedFxSelector = null;
+            this.mixer.secondPressedFxSelector = null;
+            this.mixer.resetFxSelectorColors();
+        }
+    }
+    onLongRelease() {
+        if (this.mixer.firstPressedFxSelector === null) {
+            script.toggleControl(this.group, "enabled");
+        }
+    }
+    output(enabled) {
+        if (enabled) {
+            this.send(this.color + this.brightnessOn);
+        } else {
+            // It is easy to mistake the dim state for the bright state, so turn
+            // the LED fully off.
+            this.send(this.color + this.brightnessOff);
+        }
+    }
+    presetLoaded(presetNumber) {
+        this.color = QuickEffectPresetColors[presetNumber - 2];
+        this.outConnections[1].trigger();
+    }
+    outConnect() {
+        if (this.group !== undefined) {
+            const connection0 = engine.makeConnection(this.group, "loaded_chain_preset", this.presetLoaded.bind(this));
+            if (connection0) {
+                this.outConnections[0] = connection0;
+            } else {
+                console.warn("Unable to connect '" + this.group + ".loaded_chain_preset' to the controller output. The control appears to be unavailable.");
+            }
+            const connection1 = engine.makeConnection(this.group, "enabled", this.output.bind(this));
+            if (connection1) {
+                this.outConnections[1] = connection1;
+            } else {
+                console.warn("Unable to connect '" + this.group + ".enabled' to the controller output. The control appears to be unavailable.");
+            }
+        }
+    }
+}
+
+/*
+ * Kontrol S4 Mk3 hardware-specific constants
+ */
+
+Pot.prototype.max = 2 ** 12 - 1;
+Pot.prototype.inBit = 0;
+Pot.prototype.inBitLength = 16;
+
+Encoder.prototype.inBitLength = 4;
+
+// valid range 0 - 3, but 3 makes some colors appear whitish
+Button.prototype.brightnessOff = 0;
+Button.prototype.brightnessOn = 2;
+Button.prototype.uncoloredOutput = function(value) {
+    if (this.indicatorTimer !== 0) {
+        return;
+    }
+    const color = (value > 0) ? (this.color || LedColors.white) + this.brightnessOn : LedColors.off;
+    this.send(color);
+};
+Button.prototype.colorMap = new ColorMapper({
+    0xCC0000: LedColors.red,
+    0xCC5E00: LedColors.carrot,
+    0xCC7800: LedColors.orange,
+    0xCC9200: LedColors.honey,
+
+    0xCCCC00: LedColors.yellow,
+    0x81CC00: LedColors.lime,
+    0x00CC00: LedColors.green,
+    0x00CC49: LedColors.aqua,
+
+    0x00CCCC: LedColors.celeste,
+    0x0091CC: LedColors.sky,
+    0x0000CC: LedColors.blue,
+    0xCC00CC: LedColors.purple,
+
+    0xCC0091: LedColors.fuscia,
+    0xCC0079: LedColors.magenta,
+    0xCC477E: LedColors.azalea,
+    0xCC4761: LedColors.salmon,
+
+    0xCCCCCC: LedColors.white,
+});
+
+const wheelRelativeMax = 2 ** 16 - 1;
+const wheelAbsoluteMax = 2879;
```

Comment:
Please add a comment, where this value comes from

Reply:
I have no idea. These are values that were resolved by @Be-ing, which I imagine came from the reverse engineering of the device.